src/meta/expr.lean

```diff
@@ -618,6 +618,129 @@ meta def is_eta_expansion (val : expr) : tactic (option expr) := do
 
 end expr
 
+/-! ### Declarations about `expr` manipulations -/
+
+/-- A "lens" for looking into the subterms of an expression, tracking where we've been, so that
+when we "zoom out" after making a change we know exactly which order of `congr_fun`s and
+`congr_arg`s we need to make things work.
+
+You're supposed to think of an `expr_lens` as a big set of nested applications with a single
+hole which needs to be filled, either in a function spot or argument spot. `expr_lens.fill` can
+fill this hole and turn your lens back into a real `expr`. -/
+meta inductive expr_lens
+| app_fun : expr_lens → expr → expr_lens
+| app_arg : expr_lens → expr → expr_lens
+| entire  : expr_lens
+
+namespace expr_lens
+
+/-- Inductive type with two constructors `F` and `A`,
+that represent the function-part `f` and arg-part `a` of an application `f a`. They specify the
+directions in which an `expr_lens` should zoom into an `expr`.
+
+This type is used in the development of rewriting tactics such as
+`nth_rewrite`, and `rewrite_search` (not currently in mathlib). -/
+@[derive decidable_eq, derive inhabited]
+inductive dir
+| F
+| A
+
+/-- String representation of `dir`. -/
+def dir.to_string : dir → string
+| dir.F := "F"
+| dir.A := "A"
+
+instance : has_to_string dir := ⟨dir.to_string⟩
+
+open tactic
+
+/-- Fill the function or argument hole in this lens with the given `expr`. -/
+meta def fill : expr_lens → expr → expr
+| entire        e := e
+| (app_fun l f) x := l.fill (expr.app f x)
+| (app_arg l x) f := l.fill (expr.app f x)
+
+/-- Zoom into `e : expr` given the context of an `expr_lens`, popping out an `expr` and a new
+zoomed `expr_lens`, if this is possible (`e` has to be an application). -/
+meta def zoom : expr_lens → list dir → expr → option (expr_lens × expr)
+| l [] e := (l, e)
+| l (dir.F :: rest) (expr.app f x) := (expr_lens.app_arg l x).zoom rest f
+| l (dir.A :: rest) (expr.app f x) := (expr_lens.app_fun l f).zoom rest x
+| _ _ _ := none
+
+/-- Convert an `expr_lens` into a list of instructions needed to build it; repeatedly inspecting a
+function or its argument a finite number of times. -/
+meta def to_dirs : expr_lens → list dir
+| expr_lens.entire        := []
+| (expr_lens.app_fun l _) := l.to_dirs.concat dir.A
+| (expr_lens.app_arg l _) := l.to_dirs.concat dir.F
+
+/-- Sometimes `mk_congr_arg` fails, when the function is 'superficially dependent'.
+This hack `dsimp`s the function before building the `congr_arg` expression.
+
+Unfortunately it creates some dummy hypotheses that I can't work out how to dispose of cleanly. -/
```

Comment:
What do you mean by "it creates dummy hypotheses"? Which step adds to the context, just `definev`?

Reply:
I do not know what it means. @semorrison wrote this a long time ago.